crates/findex/src/implementations/redis/findex.rs

```diff
@@ -0,0 +1,434 @@
+    ///
+    /// - `master_key`                      : master key used to generate the
+    ///   current index
+    /// - `new_master_key`                  : master key used to generate the
+    ///   new index
+    /// - `label`                           : label used to generate the new
+    ///   index
+    /// - `num_reindexing_before_full_set`  : average number of calls to compact
+    ///   needed to recompute all of the Chain Table.
+    ///
+    /// **WARNING**: the compact operation *cannot* be done concurrently with
+    /// upsert operations. This could result in corrupted indexes.
+    pub async fn compact(
+        &self,
+        master_key: &[u8; 16],
+        new_master_key: &[u8; 16],
+        label: &[u8],
+        num_reindexing_before_full_set: u32,
+    ) -> Result<(), FindexRedisError> {
+        FindexCompact::compact(
+            self,
+            &KeyingMaterial::<MASTER_KEY_LENGTH>::from(*master_key),
+            &KeyingMaterial::<MASTER_KEY_LENGTH>::from(*new_master_key),
+            &Label::from(label),
+            num_reindexing_before_full_set,
+        )
+        .await?;
+        Ok(())
+    }
+}
+
+impl FindexCallbacks<FindexRedisError, UID_LENGTH> for FindexRedis {
+    async fn progress(
+        &self,
+        _results: &HashMap<Keyword, HashSet<IndexedValue>>,
+    ) -> Result<bool, FindexRedisError> {
+        //TODO: allow passing callback fn on connect
+        Ok(true)
+    }
+
+    async fn fetch_all_entry_table_uids(&self) -> Result<Uids<UID_LENGTH>, FindexRedisError> {
+        let keys: Vec<Vec<u8>> = self.mgr.clone().keys(key(FindexTable::Entry, b"*")).await?;
+        trace!("fetch_all_entry_table_uids num keywords: {}", keys.len());
+        Ok(Uids(
+            keys.iter()
+                .map(|v| {
+                    let mut uid = [0u8; UID_LENGTH];
+                    uid.copy_from_slice(&v[4..]);
+                    Uid::from(uid)
+                })
+                .collect(),
+        ))
+    }
+
+    async fn fetch_entry_table(
+        &self,
+        entry_table_uids: Uids<UID_LENGTH>,
+    ) -> Result<EncryptedMultiTable<UID_LENGTH>, FindexRedisError> {
+        trace!(
+            "fetch_entry_table num keywords: {}:",
+            entry_table_uids.0.len(),
+        );
+        // guard against empty uids
+        if entry_table_uids.0.is_empty() {
+            return Ok(EncryptedMultiTable::default());
+        }
+
+        // build Redis keys
+        let keys: Vec<Vec<u8>> = entry_table_uids
+            .0
+            .iter()
+            .map(|uid| key(FindexTable::Entry, uid))
+            .collect();
+
+        // mget the values from the Redis keys
+        let values: Vec<Vec<u8>> = self.mgr.clone().mget(keys).await?;
+
+        // discard empty values
+        let tuples = entry_table_uids
+            .0
+            .into_iter()
+            .zip(values)
+            .filter(|(_uid, v)| !v.is_empty())
+            .collect::<Vec<_>>();
```

Comment:
I think the iterator is in arbitrary order but constant in the same run. So it's fine here (the same iteration is done above to create the Redis keys and Redis returns the values in the same order, so the second iteration should be in the same order as the first above)

Reply:
Yes, as long as it does not change between two cals. My point is that it is the case now, but there is no guarantee that this will be true in the future. But we still can change this piece of code the day it does not work anymore so this is okay.

There is still an issue with the multiple ET though. Do we ignore this use case for the Redis callback ?

Reply:
This code also relies on the assumption that the call to Redis does not shuffle the UIDs.

At the very least, these two implicit assumptions should be documented in a comment imho.

Reply:
Redis does not shuffle by design.

For the rest, not sure there is a problem but to make sure (at the cost of another iterator)

```
       // collect uids from the entry table
        let uids: Vec<Uid<32>> = entry_table_uids.0.into_iter().collect();

        // build Redis keys
        let keys: Vec<Vec<u8>> = uids
            .iter()
            .map(|uid| key(FindexTable::Entry, uid))
            .collect();

        // mget the values from the Redis keys
        let values: Vec<Vec<u8>> = self.manager.clone().mget(keys).await?;

        // discard empty values
        let tuples = uids
            .into_iter()
            .zip(values)
            .filter(|(_uid, v)| !v.is_empty())
            .collect::<Vec<_>>();
```

Same fix reported on the chain table